data/imgur.1
.\" Process this file with
.\" groff -man -Tascii foo.1
.\"
.TH IMGUR 1 "January 2011" Linux "User Manuals"
.SH NAME
imgur \- upload images to the web
.SH SYNOPSIS
.B imgur [-b]
.I filename
.br
.B imgur -l
.br
.B imgur -r [-b]
.I record_name
.SH DESCRIPTION
.B imgur
uploads images to imgur.com, and optionally
launches the system web browser to view the
uploaded image.
It can also list any previously
uploaded images, and re-display their
details.
No account or payment is needed to host images
on imgur.com, and therefore this program should
operate correctly with no configuration.

Details of each image are printed to standard
output in three tab-delimited columns. The
first column is an identifier for the image.
The second is a field name, and the third is a value.
This allows you to use
.B grep(1)
and
.B cut(1)
to pull out only the parts you need.
A list of valid field names is given in
.B KEYS
, below.
.SH OPTIONS
.IP -b
Launch the browser to look at the uploaded image.
This may be combined with the
.B -r
option, to allow you to look at a previously-uploaded
image in the browser.
.IP -l
List all previously uploaded images to standard output.
The images are listed newest first.
Their identifiers may be in the form of filenames; if so,
they represent thumbnails of the images, and only the
part without the path or extension is significant.
If there are no previously uploaded files, prints
"(no records to list)" to standard error.
.IP -r
View the records of a previously-uploaded image.
Identifiers are case-sensitive.
The identifier may be given in plain form, such as
.I LThFy
, or in the form of a filename, such as
.I /home/user/.local/share/imgur/LThFy.jpg
-- if given as a filename, the path and extension
are discarded.
You can obtain a list of valid identifiers using the
.B -l
option.
.SH KEYS
These are the strings you may find in the second column
of the results tables which are printed to standard output.
.IP delete_hash
A secret value used by the API to allow deletion of the
copy of the image on the web. Not currently usable by
this program, though it will be later.
.IP delete_page
The address of a web page which includes the
.I delete_hash
mentioned above. This allows you to delete the copy
of the image on the web using a web browser.
.IP filename
The original name of the file on your computer which was
uploaded.

This field is inserted by the program locally; it is
not generated by the remote server. As such, it is not
printed after an upload, only as part of a historical record.
.IP image_hash
A six-character identifier which represents the copy of
the image on the web. This is the simplest form which
may be used to refer to the image. It is case-sensitive.
.IP imgur_page
The address of a web page where you may view the image.
This address is what the
.B -b
option opens in the web browser.
.IP large_thumbnail
The web address of a large thumbnail of the image,
suitable for inclusion in blogs.  This image is generally
at most 640 pixels on its longest side.
.IP local_thumbnail
A local copy of the image referenced by
.I small_thumbnail
whose filename can also be used as a reference
to the image as a whole. This is the filename
which the
.B -l
option attempts to use in its listings.

This field is inserted by the program locally; it is
not generated by the remote server. As such, it is not
printed after an upload, only as part of a historical record.
.IP original_image
The web address for a copy of the original image,
at the same size.
.IP small_thumbnail
The web address of a small thumbnail of the image.
This image is generally at most 100 pixels on its longest side.
.IP time
The time at which the upload was made. This is stored
as a number of seconds since the epoch, but printed
as a human-readable date. The date format is sortable.

This field is inserted by the program locally; it is
not generated by the remote server. As such, it is not
printed after an upload, only as part of a historical record.
.SH DIAGNOSTICS
The following diagnostics may be issued on standard error:
 
(no records to list)
.RS
You requested a listing of previous uploads,
but there are none to list. Either you have never
uploaded files using this program, or the program
has been told not to log uploads.
.RE
(name) found 0
.RS
You asked to list the details of a record which
was not found.
.SH FILES
These paths are derived from the specifications given
in the XDG Base Directory Specification, and so may
vary according to the provisions of that document.
.IP /etc/xdg/imgur/imgur.conf
Site-wide configuration file, in key-file format.
Currently ignored: this is a bug.
The file is well-commented;
copy it to the path given in the
next entry to customise it for yourself.
.IP ~/.config/imgur/imgur.conf
Local configuration, in the same format
as the preceding entry.
In particular, you may set the
.I recording
key in the
.I general
group to 0 to prevent recording of uploads.
.IP ~/.local/share/imgur/uploads.conf
The details of previous uploads.
Thumbnails of these uploads are stored in the
same directory.
.SH BUGS
See
.I https://github.com/tthurman/imgur-integration/issues
for an up-to-date list of issues.

See
.I https://github.com/tthurman/imgur-integration/wiki/roadmap
to get an idea about priorities for fixing them.

Patches and suggestions are always welcome.
.SH AUTHOR
Thomas Thurman <thomas@thurman.org.uk>
.SH SEE ALSO
Further documentation may be found at
.I https://github.com/tthurman/imgur-integration/wiki


NEWS
2011-01-31 - v0.5.1
   - Don't add a launcher icon; instead, pop up help after install
     (but only on Maemo).
   - Bugfixes.

2011-01-31 - v0.5.0 - "Fox-Talbot"
   - Maemo libsharing plugin.

2011-01-27 - v0.4.0 - "Stieglitz"
   - Configuration files.
   - A man page.
   - Recording of details of previous uploads, API to check them,
     and commandline switches to access this API.
   - Details are always printed to stdout, even if the browser is
     launched.
   - Extra column in stdout details, giving the ID of the whole record.
   - More tests.
   - Bugfixes.

2011-01-22 - v0.3.0 - "Daguerre"
   - Relative pathnames may now be passed to the commandline app.
   - Some unit tests split out (but not yet made into a comprehensive
     testing system).
   - May now be built without the eog plugin.
   - libsocialweb support is now a compile-time switch, defaulting to off.
   - Added a tool to turn on the eog plugin from an installation script.

2011-01-15 - v0.2.0 - "Ansell"
   - First release.


src/cmd/imgur.c
#include <stdlib.h>
#include <stdio.h>
#include <glib.h>
#include <string.h>
#include <time.h>
#include <dbus/dbus-glib-bindings.h>
#include "../../data/imgur-client-glue.h"
#include <config.h>

#define UPLOAD_METHOD "Upload"
#define TIMEOUT_MS 3000

enum {
	EXIT_OK = 0,
	EXIT_ARGUMENT_PARSE_ERROR,
	EXIT_BUS_ERROR,
	EXIT_NO_ARGUMENT,
	EXIT_NO_RECORD_ID,
	EXIT_IMGUR_ERROR = 255
};

gboolean show_browser = FALSE;
gboolean list_all_records = FALSE;
gboolean list_some_records = FALSE;
gchar *filename = NULL;

DBusGConnection *connection = NULL;
DBusGProxy *uploader = NULL;

static GOptionEntry entries[] =
{
	{ "browser", 'b', 0, G_OPTION_ARG_NONE, &show_browser,
		"Launch browser if successful", NULL },
	{ "list", 'l', 0, G_OPTION_ARG_NONE, &list_all_records,
		"List previous uploads", NULL },
	{ "records", 'r', 0, G_OPTION_ARG_NONE, &list_some_records,
		"List details of a previous upload", NULL },
	{ NULL }
};

/**
 * Parses the commandline.
 *
 * \param argc  The number of commandline arguments, plus one.
 * \param argv  The name of the program, followed by the
 *              commandline arguments.
 */
static void
parse_commandline (int argc, char **argv)
{
	GError *error = NULL;
	GOptionContext *context =
		g_option_context_new ("[PATH] - post images to the web");

	g_option_context_add_main_entries (context,
		entries, "imgur");

	if (!g_option_context_parse (context,
		&argc, &argv, &error))
	  {
	    g_print ("%s\n", error->message);
	    g_error_free (error);
	    exit (EXIT_ARGUMENT_PARSE_ERROR);
	  }

	if (argc!=0)
	  {
	    filename = argv[1];
	  }

	if (filename && strncmp (filename, "file:", 5)==0)
	  {
	    filename += 5;
	  }

	if (!filename && !list_all_records)
	  {
	    gchar *help = g_option_context_get_help (context,
		FALSE, NULL);

	    g_print ("%s", help);
	    g_free (help);

	    exit (EXIT_NO_ARGUMENT);
	  }

	if (filename && !list_some_records)
	{
		/*
		 * At this point, "filename" is non-NULL and
		 * contains the filename we need. It's statically
		 * allocated. It may be relative. We need it
		 * to be allocated on the heap, and absolute.
		 */

		if (g_path_is_absolute (filename))
		{
			/*
			 * lovely, just what we want.
			 * Take a copy.
			 */
			filename = g_strdup (filename);
		}
		else
		{
			/* Attempt to canonicalise. */

			gchar *cwd = g_get_current_dir ();

			filename = g_build_filename (cwd,
					filename, NULL);

			g_free (cwd);
		}
	}
}

/**
 * Sets up the global variables concerned with DBus.
 */
static void
get_proxy (void)
{
	GError *error = NULL;

	connection = dbus_g_bus_get (DBUS_BUS_SESSION,
		&error);

	if (!connection)
	  {
	    g_print ("Could not connect to bus: %s\n",
		error->message);
	    g_error_free (error);
	    exit (EXIT_BUS_ERROR);
	  }

	uploader = dbus_g_proxy_new_for_name (connection,
		"com.imgur",
		"/com/imgur",
		"com.imgur");
}

/**
 * Launches the web browser. If that fails, prints an error
 * to stderr.
 *
 * \param url  The address to send the browser to.
 */
static void
launch_browser (const char* url)
{
	GError *error = NULL;

#ifdef MAEMO
	/*
	 * Maemo has its own rather weird way of doing this.
	 */
	DBusGProxy *proxy = dbus_g_proxy_new_for_name (connection,
			"com.nokia.osso_browser",
			"/com/nokia/osso_browser/request",
			"com.nokia.osso_browser");

	if (!dbus_g_proxy_call (proxy, "load_url", &error,
				G_TYPE_STRING, url,
				G_TYPE_INVALID,
				G_TYPE_INVALID))
	{
		g_warning ("Failed to spawn browser: %s", error->message);
		g_error_free (error);
	}
#else
	gchar *command_line = g_strdup_printf ("xdg-open %s",
			url);

	if (!g_spawn_command_line_async (command_line, &error))
	{
		g_warning ("Failed to spawn browser: %s", error->message);
		g_error_free (error);
		error = NULL;
	}

	g_free (command_line);
#endif
}

/**
 * Returns a time in human-readable format.
 *
 * \param clock  A decimal number of seconds since the epoch.
 * \return  The same time, expressed in human-readable format;
 *          or a copy of the same string if that string was invalid.
 *          Either way, it's the user's responsibility to free it.
 */
static gchar*
format_time (const gchar* clock)
{
	time_t time_int = atol (clock);
	struct tm time_struct;
	gchar buffer[2048];

	if (time_int==0)
	{
		return g_strdup (clock);
	}

	localtime_r (&time_int, &time_struct);

	strftime (buffer, sizeof (buffer),
		"%Y-%m-%d %H:%M:%S",
                &time_struct);

	return g_strdup (buffer);
}

/**
 * Prints one line of a hash table. Ancillary to print_hash_table().
 *
 * \param key  The key of the hash; a cast char*.
 *             If this is "imgur_page", and the global variable
 *             show_browser is TRUE, this has the side-effect of
 *             launching the browser on the URL given in value.
 * \param value  The value of the hash; a cast GValue containing
 *               a string.
 * \param identifier  An identifier for the whole hash.
 */
static void
print_one_line (gpointer key, gpointer value, gpointer identifier)
{
	GValue *v = (GValue*) value;
	const gchar *key_str = (gchar*) key;
	const gchar *value_str = g_value_get_string (value);
	gchar *temp = NULL;

	if (strcmp (key_str, "time")==0)
	{
		temp = format_time (value_str);
	}

	g_print ("%s\t%s\t%s\n",
	  identifier,
	  key_str,
	  temp? temp: value_str);

	g_free (temp);

	if (show_browser &&
		strcmp (key_str, "imgur_page")==0)
	{
		launch_browser (value_str);
	}
}

/**
 * Prints the contents of a hash table mapping
 * strings to GValues containing strings.
 * See print_one_line() for special cases.
 * If the hash is null, a dummy record mapping
 * "found" to "0" is printed.
 *
 * \param hash   The hash whose contents should
 *               be printed.
 * \param title  A title for the hash, to be
 *               printed in the first column.
 *               If this is NULL, it's guessed
 *               from the contents of the hash.
 */
static void
print_hash_table (GHashTable *hash,
	gchar *title)
{
	GError *error = NULL;
	if (!title)
	{
		/*
 		 * Oh dear. I suppose we'd better
 		 * look it up, then.
 		 */

		GValue *v = g_hash_table_lookup (hash,
			"image_hash");

		if (v)
			title = (gchar*) g_value_get_string(v);
		else
			title = "???";
	}

	/*
 	 * Now actually print stuff.
 	 */

	g_hash_table_foreach (hash,
			print_one_line,
			title);

	if (g_hash_table_size (hash)==0)
	{
		/*
 		 * No such record.  But
		 * output a dummy record
		 * so that the user knows
		 * we did actually look.
		 */

		g_print ("%s\tfound\t0\n",
			title);
	}
}

/**
 * Uploads an image, according to the commandline options.
 */
static void
perform_upload (void)
{
	GError *error = NULL;
	gchar *action = NULL;
	GHashTable *result = NULL;

	if (com_imgur_upload (uploader, filename, &result, &action, &error))
	  {

            if (action)
            {
              g_print ("imgur: action taken: %s\n", action);
            }

            print_hash_table (result, NULL);
	    g_free (action);

	    exit (EXIT_OK);
	  }
	else
	  {
	    g_print ("error\t%s\n",
		error->message);
	    g_error_free (error);

	    exit (EXIT_IMGUR_ERROR);
	  }
}

static void
dump_list_of_pairs (GPtrArray *list)
{
	int i;
	gboolean printed = FALSE;

	for (i=0; i<list->len; i++)
	{
		GValueArray *values = g_ptr_array_index (list, i);
		GValue *id_v = g_value_array_get_nth (values, 0);
		GValue *thumb_v = g_value_array_get_nth (values, 1);

		if (id_v)
		{
			const gchar *id = g_value_get_string (id_v);
			const gchar *thumb = NULL;

			if (thumb_v)
			{
				thumb = g_value_get_string (thumb_v);
			}
			else
			{
				/* FIXME: use a real placeholder image */
				thumb = "???";
			}

			g_print ("%s\t%s\n", id, thumb);
			printed = TRUE;
		}
	}

	if (!printed)
	{
		fprintf (stderr, "(no records to list)\n");
	}
}

/**
 * Prints an identifier for every known record to standard output.
 */
static void
show_all_records (void)
{
	GError *error = NULL;
	GPtrArray *list = NULL;

	if (com_imgur_list_records (uploader, &list, &error))
	{
		dump_list_of_pairs (list);
	}
	else
	{
		fprintf (stderr, "imgur: error in listing records: %s\n",
			error->message);
		g_error_free (error);
		exit (EXIT_IMGUR_ERROR);
	}
	
	/* FIXME: free list? */
}


/**
 * Prints all details of one record to standard output.
 */
static void
show_some_records (void)
{
	GError *error = NULL;
	GHashTable *hash = NULL;

	if (com_imgur_get_record (uploader, filename, &hash, &error))
	{
		print_hash_table (hash, filename);

		exit (EXIT_OK);
	}
	else
	{
		fprintf (stderr, "imgur: error in listing records: %s\n",
			error->message);
		g_error_free (error);
		exit (EXIT_IMGUR_ERROR);
	}
}
	
int
main (int argc, char **argv)
{
	g_type_init ();

	parse_commandline (argc, argv);

	get_proxy ();

	if (list_all_records)
		show_all_records ();

	if (filename)
	{
		if (list_some_records)
		{
			show_some_records ();
		}
		else
		{
			perform_upload ();
		}
	}
}

/* EOF imgur.c */

